packages/components/src/navigation/use-create-navigation-tree.js

```diff
@@ -18,6 +23,47 @@ export const useCreateNavigationTree = () => {
 		removeNode: removeMenu,
 	} = useNavigationTreeNodes();
 
+	/**
+	 * Stores direct nested menus of menus
+	 * This makes it easy to traverse menu tree
+	 *
+	 * Key is the menu prop of the menu
+	 * Value is an array of menu keys
+	 */
+	const [ directSubMenus, setDirectSubMenus ] = useState( {} );
+	const getDirectSubMenus = ( menu ) => directSubMenus[ menu ] || [];
+
+	const traverseMenu = ( startMenu, callback ) => {
+		const visited = [];
+		let queue = [ startMenu ];
+		let current;
+
+		while ( queue.length > 0 ) {
+			current = getMenu( queue.shift() );
+
+			if ( ! current || visited.includes( current.menu ) ) {
+				continue;
+			}
+
+			visited.push( current.menu );
+			queue = [ ...queue, ...getDirectSubMenus( current.menu ) ];
+
+			callback( current );
```

Comment:
I have an idea about making this a bit more performant when a menu exists

```suggestion
if ( ! callback( current ) ) {
  break;
}
```

then bellow, in `isMenuEmpty`, we can return false when we are ready to stop the algorithm. That way, we could do:

```js
let isEmpty = true;
traverseMenu( menuToCheck, ( current ) => {
    if ( ! current.isEmpty ) {
	    isEmpty = false;
	}
    // Continue until a menu has items.
    return current.isEmpty;
} );
return isEmpty;
```

Reply:
Great idea! Thanks! Implemented it.